Source/UROSWorldControl/Public/Relocator.h

```diff
@@ -0,0 +1,80 @@
+// Fill out your copyright notice in the Description page of Project Settings.
+
+#pragma once
+#include "CoreMinimal.h"
+#include "GameFramework/Actor.h"
+#include "ROSBridgeHandler.h"
+#include "ROSBridgeSrvServer.h"
+#include "ROSBridgeSrv.h"
+#include "Engine/StaticMeshActor.h"
+#include "UTags.h"
+#include "TagStatics.h"
+#include "ROSWorldControlManager.h"
+#include "Relocator.generated.h"
+
+struct MoveAssetParams
+{
+	AActor* Actor;
+	FVector Location;
+	FRotator Rotator;
+};
+
+UCLASS()
+class UROSWORLDCONTROL_API ARelocator : public AActor
+{
+	GENERATED_BODY()
+
+public:
+	// Sets default values for this actor's properties
+	ARelocator();
+
+	TArray<MoveAssetParams> MoveAtNextTick;
+	TSharedPtr<FROSBridgeHandler> Handler;
+	AROSWorldControlManager* Controller;
+
+protected:
+	// Called when the game starts or when spawned
+	virtual void BeginPlay() override;
+
+public:
+	// Called every frame
+	virtual void Tick(float DeltaTime) override;
+	bool Relocate(AActor* Actor, FVector Location, FRotator Rotator);
+	UPROPERTY(EditAnywhere, Category = "RosBridge Websocket")
+		FString ServerAdress = TEXT("192.168.1.19");
+
+	UPROPERTY(EditAnywhere, Category = "RosBridge Websocket")
+		int ServerPort = 9090;
+
+
+	UPROPERTY(EditAnywhere, Category = "ROS")
+		FString NameSpace = TEXT("unreal");
+
+	virtual void EndPlay(const EEndPlayReason::Type Reason);
+
+
+private:
+
+	class FROSRelocationServer final : public FROSBridgeSrvServer
+	{
+	private:
+		ARelocator * Parent;
+		bool GameThreadDoneFlag;
+		bool ServiceSuccess;
+
+	public:
+		FROSRelocationServer(FString NameSpace, FString Name, ARelocator* Parent_) :
+			FROSBridgeSrvServer(NameSpace + TEXT("/") + Name, TEXT("unreal_msgs/relocate_model"))
+		{
+			Parent = Parent_;
+		}
+
+		TSharedPtr<FROSBridgeSrv::SrvRequest> FromJson(TSharedPtr<FJsonObject> JsonObject) const override;
+
+		TSharedPtr<FROSBridgeSrv::SrvResponse> Callback(TSharedPtr<FROSBridgeSrv::SrvRequest> Request) override;
+
+		void SetGameThreadDoneFlag(bool Flag);
```

Comment:
Is this threadsafe?

Reply:
It is not, if the Handler is running two callbacks on the same server object at the same time. However that shouldn't be the case, since then every line of the callback code would run into trouble with multi threading.
If the Callback is only called once it is safe, since the this is essentially just a "join" statement. The flag can only be Set to `false` by the callback function, and will wait until its `true` to terminate. The AsyncTask is the only one setting it `true`, and since its not possible for two callbacks to run at the same time it's not possible that there are two AsyncTask manipulating the same ServerObject. This is the case because the lifetime of the AsyncTask is always smaller than the lifetime of the Callback, since it spawns the Task and waits for it.

I also made sure the actual return value is set before the Flag is, since this could otherwise mean trouble under the right thread circumstances.